analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -51,30 +51,31 @@ public abstract class UtilityAnalyzerBase : SonarDiagnosticAnalyzer
                 EndOffset = lineSpan.EndLinePosition.Character
             };
 
-        protected void ReadParameters(AnalyzerOptions options, string language)
+        protected void ReadParameters(SonarAnalysisContext context, CompilationAnalysisContext c)
         {
-            var settings = PropertiesHelper.GetSettings(options);
-            var projectOutputAdditionalFile = options.AdditionalFiles.FirstOrDefault(IsProjectOutput);
-
-            if (!settings.Any() || projectOutputAdditionalFile == null)
+            var settings = PropertiesHelper.GetSettings(c.Options);
+            var outPath = context.ProjectConfiguration(c.Options).OutPath;
+            // For backward compatibility with S4MSB <= 5.0
+            if (outPath == null && c.Options.AdditionalFiles.FirstOrDefault(IsProjectOutFolderPath) is { } projectOutFolderAdditionalFile)
             {
-                return;
+                outPath = File.ReadAllLines(projectOutFolderAdditionalFile.Path).First();
             }
-
-            lock (parameterReadLock)
+            if (settings.Any() && outPath != null)
             {
-                IgnoreHeaderComments = PropertiesHelper.ReadIgnoreHeaderCommentsProperty(settings, language);
-                AnalyzeGeneratedCode = PropertiesHelper.ReadAnalyzeGeneratedCodeProperty(settings, language);
-                OutPath = File.ReadAllLines(projectOutputAdditionalFile.Path).FirstOrDefault(l => !string.IsNullOrEmpty(l));
-                if (!string.IsNullOrEmpty(OutPath))
+                lock (parameterReadLock)
                 {
-                    OutPath = Path.Combine(OutPath, language == LanguageNames.CSharp ? "output-cs" : "output-vbnet");
-                    IsAnalyzerEnabled = true;
+                    IgnoreHeaderComments = PropertiesHelper.ReadIgnoreHeaderCommentsProperty(settings, c.Compilation.Language);
+                    AnalyzeGeneratedCode = PropertiesHelper.ReadAnalyzeGeneratedCodeProperty(settings, c.Compilation.Language);
+                    if (!string.IsNullOrEmpty(outPath))
```

Comment:
you are checking outPath for null twice...

if between the first `if` and the `lock`, the value for `outPath` can change, that means that the value for settings can change, too - so both should be checked again inside the `lock`, right?

Also, could you please use the same check in line 63 and line 69 (`IsNullOrEmpty`), otherwise, it makes the code a bit confusing

Reply:
This is actually a nice catch that I missed and that pop up after deobfuscation.

Reply:
While looking at it, I don't actually see a reason for the `lock` itself as well.

Reply:
Yes, it's a read-only operation. 

Even more (and that's for a separate PR/issue, but now we're in the context) - I'm not sure the `FileWriteLock` in the generic class is needed. It's a compilation end method. And each analyzer writes to its specific protobuf file. And we don't expect multiple analyzers to run in parallel.

And regarding this class, we have a performance hotspot in [TokenClassifierBase](https://github.com/SonarSource/sonar-dotnet/blob/b8ec1a6e30f781b949941b92542f7088c39b676f/analyzers/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs#L67) because we query the SemanticModel for each token, so we would probably improve a lot with some caching. But that's another topic.